### PR TITLE
Fix various bugs for "install" page

### LINKED ISSUE #20089: During new instance setup make 'Gitea Base URL' filled from `window.location.origin`
### Feature Description

During new instance setup make 'Gitea Base URL' filled from `window.location.origin`.

### LINKED ISSUE #23183: GET data: net::ERR_INVALID_URL
### Description

Delete `data` and `app.ini`, run the server, access the install page, and you will get this error in console
image

~~This will not happen in 1.18.x~~

### Gitea Version

latest

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

localhost

### Database

None

### LINKED ISSUE #23184: Redirect failed after install gitea
### Description

After you install the gitea, it will redirect to user/login,
In 1.18 and 1.19, it works well but in the latest version,
you will get an error in console:
image

### Gitea Version

latest

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

devcontainer

### Database

None

### PR DESCRIPTION
**Pull request by wxiaoguang**
## TLDR

* Fix the broken page / broken image problem when click "Install"
* Close #20089
* Fix the Password Hash Algorithm display problem for #22942
* Close #23183
* Close #23184

### The broken page / broken image problem when click "Install" (Redirect failed after install gitea #23184)

Before: when click "install", all new requests will fail, because the server has been restarted. Users just see a broken page with broken images, sometimes the server is not ready but the user would have been redirect to "/user/login" page, then the users see a new broken page (connection refused or something wrong ...)

After: only check InstallLock=true for necessary handlers, and sleep for a while before restarting the server, then the browser has enough time to load the "post-install" page. And there is a script to check whether "/user/login" is ready, the user will only be redirected to the login page when the server is ready.

### During new instance setup make 'Gitea Base URL' filled from window.location.origin #20089

If the "app_url" input contains `localhost` (the default value from config), use current window's location href as the `app_url` (aka ROOT_URL)

### Fix the Password Hash Algorithm display problem for "Provide the ability to set password hash algorithm parameters #22942"

Before: the UI shows `pbkdf2$50000$50`

<details>

image

</details>

After: the UI shows `pbkdf2`

<details>

image

</details>

### GET data: net::ERR_INVALID_URL #23183

Cause by empty `data:` in `<link rel="manifest" href="data:{{.ManifestData}}">`

### CODE REVIEW

File: routers/install/routes.go [hunk 1]
```diff
@@ -107,14 +108,20 @@ func Routes(ctx goctx.Context) *web.Route {
...
 }
 
 func installNotFound(w http.ResponseWriter, req *http.Request) {
-	http.Redirect(w, req, setting.AppURL, http.StatusFound)
+	w.Header().Add("Content-Type", "text/html; charset=utf-8")
+	w.Header().Add("Refresh", fmt.Sprintf("1; url=%s", setting.AppSubURL+"/"))
+	// do not use 30x status, because the "post-install" page needs to use 404/200 to detect if Gitea has been installed.
+	// the fetch API could follow 30x requests to the page with 200 status.
+	w.WriteHeader(http.StatusNotFound)
+	_, _ = w.Write([]byte(fmt.Sprintf(`Not Found. <a href="%s">Go to default page</a>.`, html.EscapeString(setting.AppSubURL+"/"))))
```
Comment (wolfogre) [inline routers/install/routes.go]: Why not
```suggestion
	_, _ = fmt.Fprintf(w, `Not Found. <a href="%s">Go to default page</a>.`, html.EscapeString(setting.AppSubURL+"/"))
```

Comment (wxiaoguang) [reply routers/install/routes.go]: Forgot that ....

**Review by wolfogre — APPROVED** [review]

**Review by lunny — APPROVED** [review]

File: web_src/js/features/install.js [hunk 2]
```diff
@@ -40,6 +48,18 @@ export function initInstall() {
     } // else: for SQLite3, the default path is always prepared by backend code (setting)
   }).trigger('change');
 
+  const $appUrl = $('#app_url');
+  const configAppUrl = $appUrl.val();
+  if (configAppUrl.includes('://localhost') || configAppUrl.includes('://127.0.0.1')) {
```
Comment (silverwind) [inline web_src/js/features/install.js]: Missing IPv6 localhost `::1`

Comment (wxiaoguang) [reply web_src/js/features/install.js]: IMO seldom people would do so, and the default value from config is just `localhost` (actually we even do not need to care about `127.0.0.1`, I just put it there for no strong reason).

Comment (silverwind) [inline web_src/js/features/install.js]:
```suggestion
  if (configAppUrl.includes('://localhost') || configAppUrl.includes('://127.0.0.1') || configAppUrl.includes('://[::1]')) {
```

File: web_src/js/features/install.js [hunk 3]
```diff
@@ -40,6 +48,18 @@ export function initInstall() {
...
+    $appUrl.val(window.location.href);
+  }
+
+  const $domain = $('#domain');
+  const configDomain = $domain.val().trim();
+  if (configDomain === 'localhost' || configDomain === '127.0.0.1') {
```
Comment (silverwind) [inline web_src/js/features/install.js]:
```suggestion
  if (configDomain === 'localhost' || configDomain === '127.0.0.1' || configDomain === '::1') {
```

File: web_src/js/features/install.js (hunk 2 above)
Comment (silverwind) [reply web_src/js/features/install.js]: I can't accept such an answer. It's just ignorant to ignore IPv6 in 2023.

Comment (wxiaoguang) [reply web_src/js/features/install.js]: Then let's only keep the localhost, because `Domain = sec.Key("DOMAIN").MustString("localhost")`

Comment (silverwind) [reply web_src/js/features/install.js]: Could just have a little helper function. It would be useful in other places as well I suppose:

````js
function isLocalhost(hostname) {
  if (typeof hostname !== 'string') return false;
  if (hostname === 'localhost') return true;
  if (hostname === '::1') return true;
  if (/^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/.test(hostname)) return true;
  return false;
}
````

Comment (wxiaoguang) [reply web_src/js/features/install.js]: Really no necessary to do so .... because the default domain is localhost, the default app_url is `http://localhost:3000`
